pkg/dataplane/nginx.go

```diff
@@ -0,0 +1,284 @@
+	klog.InfoS("Starting NGINX Ingress configurer")
+
+	cmd := n.command.ExecCommand()
+
+	// put NGINX in another process group to prevent it
+	// to receive signals meant for the controller
+	cmd.SysProcAttr = &syscall.SysProcAttr{
+		Setpgid: true,
+		Pgid:    0,
+	}
+
+	if n.cfg.EnableSSLPassthrough {
+		n.setupSSLProxy()
+	}
+
+	klog.InfoS("Starting NGINX process")
+	n.start(cmd)
+
+	// TODO: Implement a "sync / trigger configuration queue to force full ingress resync"
+	n.syncQueue.EnqueueTask(task.GetDummyObject("initial-sync"))
+
+	// In case of error the temporal configuration file will
+	// be available up to five minutes after the error
+	go func() {
+		for {
+			time.Sleep(5 * time.Minute)
+			err := cleanTempNginxCfg()
+			if err != nil {
+				klog.ErrorS(err, "Unexpected error removing temporal configuration files")
+			}
+		}
+	}()
+
+	for {
+		select {
+		case err := <-n.ngxErrCh:
+			if n.isShuttingDown {
+				return
+			}
+
+			// if the nginx master process dies, the workers continue to process requests
+			// until the failure of the configured livenessProbe and restart of the pod.
+			if process.IsRespawnIfRequired(err) {
+				return
+			}
+
+		// TODO: Implement below, which is the main loop to resync / configure NGINX. This is the part that will watch
+		// gRPC events and trigger NGINX reload or not
+		/*
+			case event := <-n.updateCh.Out():
+				if n.isShuttingDown {
+					break
+				}
+
+				if evt, ok := event.(store.Event); ok {
+					klog.V(3).InfoS("Event received", "type", evt.Type, "object", evt.Obj)
+					if evt.Type == store.ConfigurationEvent {
+						// TODO: is this necessary? Consider removing this special case
+						n.syncQueue.EnqueueTask(task.GetDummyObject("configmap-change"))
+						continue
+					}
+
+					n.syncQueue.EnqueueSkippableTask(evt.Obj)
+				} else {
+					klog.Warningf("Unexpected event type received %T", event)
+				}*/
+		case <-n.stopCh:
+			return
+		}
+	}
+}
+
+// Stop gracefully stops the NGINX master process.
+func (n *NGINXConfigurer) Stop() error {
+	n.isShuttingDown = true
+
+	n.stopLock.Lock()
+	defer n.stopLock.Unlock()
+
+	time.Sleep(time.Duration(n.cfg.ShutdownGracePeriod) * time.Second)
+
+	// send stop signal to NGINX
+	klog.InfoS("Stopping NGINX process")
+	cmd := n.command.ExecCommand("-s", "quit")
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	err := cmd.Run()
+	if err != nil {
+		return err
+	}
+
+	// wait for the NGINX process to terminate
+	timer := time.NewTicker(time.Second * 1)
+	for range timer.C {
+		if !nginx.IsRunning() {
+			klog.InfoS("NGINX process has stopped")
+			timer.Stop()
+			break
+		}
+	}
+
+	return nil
+}
+
+func (n *NGINXConfigurer) start(cmd *exec.Cmd) {
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	if err := cmd.Start(); err != nil {
+		klog.Fatalf("NGINX error: %v", err)
+		n.ngxErrCh <- err
+		return
+	}
+
+	go func() {
+		n.ngxErrCh <- cmd.Wait()
+	}()
+}
+
+// TODO: Shall we test this in both places? (CP and DP)
```

Comment:
In fact, I think we can simplify this model.
The CP is the only source, if checking is done on the CP side, then multiple DP instances will not need to check, which can reduce costs.

Reply:
In addition, if it is checked on the DP side, when the configuration file is wrong, the behavior of the DP is to continue to run without any prompt? that error may be masked